lib/IRGen/IRGenModule.cpp

```diff
@@ -1185,64 +1175,197 @@ static bool isFirstObjectFileInModule(IRGenModule &IGM) {
   return containingModule->getFiles().front() == file;
 }
 
-void IRGenModule::emitAutolinkInfo() {
-  // Collect the linker options already in the module (from ClangCodeGen).
+static bool
+doesTargetAutolinkUsingAutolinkExtract(const SwiftTargetInfo &TargetInfo,
+                                       const llvm::Triple &Triple) {
+  if (TargetInfo.OutputObjectFormat == llvm::Triple::ELF && !Triple.isPS4())
+    return true;
+
+  if (TargetInfo.OutputObjectFormat == llvm::Triple::Wasm)
+    return true;
+
+  if (Triple.isOSCygMing())
+    return true;
+
+  return false;
+}
+
+namespace {
+
+struct AutolinkKind {
+  enum ValueTy {
+    LLVMLinkerOptions,
+    LLVMDependentLibraries,
+    SwiftAutoLinkExtract,
+  };
+
+  ValueTy Value;
+
+  AutolinkKind(ValueTy value) : Value(value) {}
+  AutolinkKind(const AutolinkKind &kind) : Value(kind.Value) {}
+
+  StringRef getSectionNameMetadata();
+
+  template <typename Vector, typename Set>
+  void collectEntriesFromLibraries(llvm::SetVector<llvm::MDNode *, Vector, Set> &Entries,
+                                   ArrayRef<LinkLibrary> AutolinkEntries,
+                                   IRGenModule &IGM);
+
+  template <typename Vector, typename Set>
+  void writeEntries(llvm::SetVector<llvm::MDNode *, Vector, Set> Entries,
+                    llvm::NamedMDNode *Metadata, IRGenModule &IGM);
+
+  static AutolinkKind create(const SwiftTargetInfo &TargetInfo,
+                             llvm::Triple Triple, IRGenLLVMLTOKind LLVMLTOKind);
+};
+} // namespace
```

Comment:
Can you put a newline in between `};` and the `}`. Can you change // namespace -> // anonymous namespace.?

Reply:
Updated 👍